### ISSUE TITLE
Could not load type 'GrainInterfaces.DerivedClassWithGenericConstraints'

### ISSUE THREAD

**Issue by peter-perot** [dotnet/orleans#1480]
Hi,

here I have an issue with a special class which causes **TypeLoadExceptions in the silo**.

How to reconstruct the bug?

Create a solution with 4 projects.

Project 1: **A simple class library** with a single class. You have to reference the Orleans.dll.

``` csharp
using Orleans;

namespace ClassLibrary
{
    public class ClassWithGenericConstraints
    {
        private static void RegisterInternal<TGrainInterface>() where TGrainInterface : IGrainWithIntegerKey
        {
        }
    }
}
```

Project 2: The **grain interface assembly**.

``` csharp
using Orleans;

namespace GrainInterfaces
{
    /// <summary>
    /// Grain interface IGrain1
    /// </summary>
    public interface IGrain1 : IGrainWithGuidKey
    {
    }
}
```

``` csharp
using ClassLibrary;

namespace GrainInterfaces
{
    public class DerivedClassWithGenericConstraints : ClassWithGenericConstraints
    {
    }
}
```

Project 3: The **grain class assembly**.

``` csharp
using GrainInterfaces;
using Orleans;

namespace GrainsClasses
{
    /// <summary>
    /// Grain implementation class Grain1.
    /// </summary>
    public class Grain1 : Grain, IGrain1
    {
    }
}
```

Project 4: The **silo host** (taken from the template).

Attached solution:
[OrleansBugDemo.zip](https://github.com/dotnet/orleans/files/137254/OrleansBugDemo.zip)

All projects compile into the same bin-folder.

Now start the silo. You get some errors/warnings.

[ICOWS566-2016-02-19-01.13.08.542Z.log.txt](https://github.com/dotnet/orleans/files/137265/ICOWS566-2016-02-19-01.13.08.542Z.log.txt)

[Client-2016-02-19-01.13.19.615Z.log.txt](https://github.com/dotnet/orleans/files/137264/Client-2016-02-19-01.13.19.615Z.log.txt)

Now delete the class _DerivedClassWithGenericConstraints_ and restart the silo. Everything works fine.

**I'm using Orleans in a larger project, and here the issue is much more complex:** I have an attribute with such a constrained generic method. These attribute is used to mark some grain interfaces. Everything works okay, but as soon as I put a class with the _Serializable_ attribute into the **grain class project**, I get a similar error/warning. When I remove the _Serializable_ attribute, everything works fine.

Maybe there is something going wrong with Roslyn code generation?

**Comment by ReubenBond** [comment]
The issue here seems to be that the assembly containing `ClassWithGenericConstraints` is not loaded by the time code generation is performed. The `RegisterInternal` method & hence generic constraint are not required to repro this.

This is not directly a code generation issue, but an issue in the supporting infrastructure. I'll see what the best way to fix it is.

In the meantime, as a workaround, you can eagerly load all of your assemblies before starting Orleans,. You can do ensure the assembly is loaded by referencing a type within it in an expression, eg: `typeof(ClassWithGenericConstraints)`.

**Comment by ReubenBond** [comment]
@peter-perot could you please submit a repro test for this? If you want to know what that looks like, see #1425 as an example.

**Comment by peter-perot** [comment]
@ReubenBond: I'm just trying to get familiar with Git. I forked the original repo, created a branch in my clone ("Issue_1480"), and pushed it back. Now I don't know how you can pull and merge my delta out of my forked repo.

https://github.com/peter-perot/orleans/commit/d5c2cec2f9edff11d8841f463c2a150de388a6b0

However: The Unit test is slighly strange. When I try to serialize 'ClassReferencingOrleansTypeDto', the unit tests comes back with a time-out after 30 secs. In my standalone application a serialization exceptions occures instead.

**Comment by peter-perot** [comment]
Update: Oops, the file containing class 'ClassNotReferencingOrleansTypeDto' is wrongly named 'ClassReferencingOrleansTypeDto.cs'. Could you please rename it to 'ClassNotReferencingOrleansTypeDto.cs' during merge?

**Comment by peter-perot** [comment]
@ReubenBond: I merged the current head of the master-branch into my repro test and filed a new pull request.

**Comment by peter-perot** [comment]
@ReubenBond: Do you have any new findings about what causes this issue?

Yesterday I discovered the same issue on client side after calling `GrainClient.Initialize(...)`.

I discovered when all assemblies which reference a "Orleans..." assembly are loaded eagerly before silo startup or grain client initialization.

Since I'm using MEF in connection with Unity IoC there are also some assemblies loaded which are not directly referenced in the C# project.

The following workaround helps me to get around this issue:

````csharp
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel.Composition.Hosting;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.Remoting;

namespace SiloHosting
{
	/// <summary>
	///   TODO: This is a workaround for an Orleans bug which occasionally occurs during runtime code generation.
	/// </summary>
	internal static class EagerLoader
	{
		public static void Execute()
		{
			var appDomain = AppDomain.CreateDomain("EagerLoaderDomain");
			ReadOnlyCollection<AssemblyName> assemblyNames;

			try
			{
				var re =
					(RemoteExecutor)
						appDomain.CreateInstanceAndUnwrap(typeof(RemoteExecutor).Assembly.FullName, typeof(RemoteExecutor).FullName);

				assemblyNames = re.GetMarkedAssemblyNames();
			}
			finally
			{
				AppDomain.Unload(appDomain);
			}

			foreach (var assemblyName in assemblyNames)
			{
				AppDomain.CurrentDomain.Load(assemblyName);
			}
		}

		internal abstract class CrossAppDomainObject : MarshalByRefObject, IDisposable
		{
			private bool isDisposed;

			~CrossAppDomainObject()
			{
				this.Dispose(false);
			}

			public static void Disconnect(MarshalByRefObject obj)
			{
				RemotingServices.Disconnect(obj);
			}

			public void Dispose()
			{
				this.Dispose(true);
				GC.SuppressFinalize(this);
			}

			public sealed override object InitializeLifetimeService()
			{
				return null;
			}

			protected virtual void Dispose(bool disposing)
			{
				if (!this.isDisposed)
				{
					Disconnect(this);
					this.isDisposed = true;
				}
			}
		}

		private sealed class RemoteExecutor : CrossAppDomainObject
		{
			public ReadOnlyCollection<AssemblyName> GetMarkedAssemblyNames()
			{
				var folders = new HashSet<string>();
				GetFoldersRecursively(GetAssemblyDirectory(typeof(RemoteExecutor).Assembly), folders);

				var files = new HashSet<string>();

				foreach (string folder in folders)
				{
					using (var cat = new DirectoryCatalog(folder))
						files.UnionWith(cat.LoadedFiles);
				}

				return Array.AsReadOnly((from loadedFile in files
					select Assembly.ReflectionOnlyLoadFrom(loadedFile)
					into assembly
					let assemblyName = assembly.GetName()
					where
						!assemblyName.Name.StartsWith("Orleans", StringComparison.Ordinal) &&
							assembly.GetReferencedAssemblies().Select(x => x.Name).FirstOrDefault(x => x == "Orleans") != null
					select assemblyName).ToArray());
			}

			private static string GetAssemblyDirectory(Assembly assembly)
			{
				if (assembly == null)
				{
					throw new ArgumentNullException(nameof(assembly));
				}

				string codeBase = assembly.CodeBase;
				var uri = new UriBuilder(codeBase);
				string path = Uri.UnescapeDataString(uri.Path);
				return Path.GetDirectoryName(path);
			}

			private static void GetFoldersRecursively(string path, HashSet<string> folders)
			{
				if (!folders.Add(path))
				{
					return;
				}

				foreach (string child in Directory.GetDirectories(path))
				{
					GetFoldersRecursively(child, folders);
				}
			}
		}
	}
}
````

Simply calling `EagerLoader.Execute()` first loads all assemblies in a temporary AppDomain just for analyzing them. Then it returns the names of all assemblies which reference an "Orleans*" assembly (any assembly the name of which starts with "Orleans". Now these assemblies are loaded into the actual AppDomain.

**Comment by ReubenBond** [comment]
@peter-perot do you need to use a separate AppDomain? What happens if you eagerly load the assemblies in the main AppDomain?

**Comment by peter-perot** [comment]
@ReubenBond: The reason for using a separate (temporary) AppDomain is that the utility scans **all** assemblies in the silo directory recursively in order to gather the assembly names of the assemblies which have to be loaded eagerly (but from the primary AppDomain). This way I can free memory by unloading the unused assemblies.

The first version of that utility did not use a second AppDomain. It simply loaded all assemblies into the primary AppDomain, and it worked, too. But this way all unused (but scanned) assemblies have been locked into memory since assemblies cannot be unloaded from an AppDomain separately.

**Comment by sergeybykov** [comment]
Closing this, as we don't scan all assemblies by default in 2.0.0.
